pkg/aggregator/demultiplexer.go

```diff
@@ -393,52 +459,143 @@ func (d *AgentDemultiplexer) Stop(flush bool) {
+		blockChan:         make(chan struct{}),
+	}
+	d.flushChan <- trigger
+	<-trigger.blockChan
 }
 
-// ForceFlushToSerializer flushes all data from the aggregator and time samplers
+// flushToSerializer flushes all data from the aggregator and time samplers
 // to the serializer.
-// Not safe to call from multiple threads, consider using ForceFlushToSerializer instead.
-func (d *AgentDemultiplexer) forceFlushToSerializer(start time.Time, waitForSerializer bool) {
+//
+// Best practice is that this method is *only* called by the flushLoop routine.
+// It technically works if called from outside of this routine, but beware of
+// deadlocks with the parallel stream series implementation.
+//
+// This implementation is not flushing the TimeSampler and the BufferedAggregator
+// concurrently because the IterableSeries is not thread safe / supporting concurrent usage.
+// If one day a better (faster?) solution is needed, we could either consider:
+// - to have an implementation of SendIterableSeries listening on multiple sinks in parallel, or,
+// - to have a thread-safe implementation of the underlying `util.BufferedChan`.
+func (d *AgentDemultiplexer) flushToSerializer(start time.Time, waitForSerializer bool) {
+	d.m.Lock()
+	defer d.m.Unlock()
 
-	// flush the time samplers
-	// ----------------------
+	if d.aggregator == nil {
+		// NOTE(remy): we could consider flushing only the time samplers
+		return
+	}
 
-	if waitForSerializer {
-		wg := sync.WaitGroup{}
-		for _, tsWorker := range d.statsd.workers {
-			wg.Add(1)
-			// order the flush to the time sampler, and wait, in a different routine
-			go func(worker *timeSamplerWorker, wg *sync.WaitGroup) {
-				worker.flush(start, true)
-				wg.Done()
-			}(tsWorker, &wg)
-		}
-		// wait for all samplers to have finished their flush
-		wg.Wait()
-	} else {
-		for _, tsWorker := range d.statsd.workers {
-			tsWorker.flush(start, false)
+	flushedSeries := make([]metrics.Series, 0)
+	flushedSketches := make([]metrics.SketchSeriesList, 0)
+
+	// only used when we're using flush/serialize in parallel feature
+	var seriesSink *metrics.IterableSeries
+	var done chan struct{}
+
+	if d.aggregator.flushAndSerializeInParallel.enabled {
+		seriesSink = metrics.NewIterableSeries(func(se *metrics.Serie) {
+			tagsetTlm.updateHugeSerieTelemetry(se)
+		}, d.aggregator.flushAndSerializeInParallel.bufferSize, d.aggregator.flushAndSerializeInParallel.channelSize)
+		done = make(chan struct{})
+		go d.sendIterableSeries(start, seriesSink, done)
+	}
+
+	// flush DogStatsD pipelines (statsd/time samplers)
+	// ------------------------------------------------
+
+	for _, worker := range d.statsd.workers {
+		// order the flush to the time sampler, and wait, in a different routine
+		trigger := flushTrigger{
+			time:            start,
+			blockChan:       make(chan struct{}),
+			flushedSeries:   &flushedSeries,
+			flushedSketches: &flushedSketches,
+			seriesSink:      seriesSink,
 		}
+
+		worker.flushChan <- trigger
```

Comment:
I don't think we want to abstract this more than this, especially since it is tightly close to the `Demultiplexer` implementation (we don't want something else to flush those?). But I can get behind an extra function call if you really want this behind a function. Let me know.